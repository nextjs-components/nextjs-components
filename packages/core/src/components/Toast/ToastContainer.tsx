import clsx from "clsx";
import React from "react";
import { useCallback, useEffect, useReducer, useRef, useState } from "react";

import { Button } from "../Button";
import type { IToast } from "./ToastsProvider";
import styles from "./toasts.module.css";

interface ToastContainerProps extends IToast {
  height: number;
  heights?: number[];
  hovering: boolean;
  onMount?: () => void;
  position?: number;
  preserve?: boolean;
  remove?: () => void;
  text: string;
  id: string;
}

const ToastContainer: React.ComponentType<ToastContainerProps> = ({
  // children,
  text,
  position,
  height,
  heights,
  onMount,
  remove,
  preserve,
  type,
  action,
  cancelAction,
  hovering,
  ...props
}) => {
  const [hiding] = useState(false);
  const [visible, setVisible] = useState(false);

  useEffect(() => {
    setTimeout(() => setVisible(true), 0);
  }, []);

  useEffect(() => {
    // force update, because...
    // the heights array (mutable ref) gets mutate
    // we want the changes to be repainted
    setTimeout(() => onMount?.(), 0);
  }, [onMount]);

  useEffect(() => {
    // auto-remove, but don't remove if hovering
    let id: NodeJS.Timeout;

    if (!preserve && !hovering) {
      id = setTimeout(remove, 4000);
    }
    return () => {
      clearTimeout(id);
    };
  }, [hovering, preserve]);

  // get the height of the toast
  const heightRef = useRef(height);
  // mutate the array

  useEffect(() => {
    // when mounted, update the heights array with current height
    heights[position] = heightRef.current;
    // when unmounting, remove current height from the heights array
    return () => {
      heights.splice(position, 1);
    };
  }, [position, heights]);

  const sum = useCallback((a: number, b: number) => a + b, []);

  return (
    <div
      data-geist-toast=""
      className={clsx("toast-container", styles.toastContainer, {
        [styles.visible]: visible,
        [styles[type]]: !!type,
      })}
      style={{
        maxHeight: position === 0 ? heightRef.current : 50,
        transform:
          // prettier-ignore
          position === 0
            ? visible && "none"
            : `translate3d(0px, calc(-${heights[0]}px + 100% + -${20 * position}px),-${position}px)
               scale(${1 - position * 0.05})`,
      }}
    >
      <div className={styles.toast}>
        <div // inner text
          ref={(el) => (heightRef.current = el?.offsetHeight - 1 + 48)}
          className={clsx(styles.message, {
            [styles.action]: action,
            [styles.cancel]: cancelAction,
          })}
        >
          {text}
        </div>
        {cancelAction && (
          <Button
            size="small"
            type="secondary"
            style={{ marginRight: 10 }}
            className={clsx("button", "small")}
          >
            {cancelAction}
          </Button>
        )}
        {action && (
          <Button size="small" className={clsx("button", "small")}>
            {action}
          </Button>
        )}
      </div>
      <style jsx>{`
        :global(.toast-area:hover) .toast-container {
          max-height: ${heightRef.current}px !important;
          transform: translate3d(
              0px,
              -${heights.slice(0, position).reduce(sum, 0) + position * 20}px,
              ${position * -1}px
            )
            scale(1) !important;
        }
        :global(.button.small) {
          min-width: auto;
          height: 24px;
          line-height: 22px;
          padding: 0 10px 0 10px;
        }
      `}</style>
    </div>
  );
};

export default ToastContainer;

// Multiline
// translate3d(0px, -236px, -2px)
//  - height 98
// translate3d(0px, -118px, -1px) scale(1) !important;
//  - height 98
// translate3d(0px, 0px, 0px) scale(1) !important;
//  - height 98

// ransform: translate3d( 0px, 118px, --1px ) scale(1) !important;
