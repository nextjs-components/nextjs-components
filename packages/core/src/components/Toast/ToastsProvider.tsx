import React from "react";
import { createContext, useRef, useState } from "react";

type ToastType = "success" | "error" | "violet";
export interface IToast {
  height: number;
  key: string;
  preserve?: boolean;
  text: string;
  type?: ToastType;
  action?: string;
  cancelAction?: string;
}

export interface IToastsContext {
  current?: {
    messages: IToast[];
    message: (message: Partial<IToast> | string) => void;
    error: (message: Partial<IToast> | string) => void;
    success: (message: Partial<IToast> | string) => void;
    removeToastByKey: (key: string) => void;
  };
}

const throwerFn = () => {
  throw new Error("Missing a ToastsProvider");
};

export const ToastsContext = createContext<IToastsContext>({
  current: {
    messages: [],
    message: throwerFn,
    error: throwerFn,
    success: throwerFn,
    removeToastByKey: throwerFn,
  },
});

const ToastsProvider = ({ children }) => {
  const [messages, setMessages] = useState<IToast[]>([]);

  const message = (toast: Partial<IToast> | string) => {
    let text, preserve, action, cancelAction;
    if (typeof toast === "string") {
      text = toast;
    } else {
      ({ text, preserve, action, cancelAction } = toast);
    }
    setMessages((prev) => [
      ...prev,
      {
        text,
        preserve,
        height: action || cancelAction ? 72 : 64,
        key: (+new Date()).toString(),
        action,
        cancelAction,
      },
    ]);
  };
  const error = (toast: Partial<IToast> | string) => {
    let text, preserve, action, cancelAction;
    if (typeof toast === "string") {
      text = toast;
    } else {
      ({ text, preserve, action, cancelAction } = toast);
    }
    setMessages((prev) => [
      ...prev,
      {
        text,
        preserve,
        height: action || cancelAction ? 72 : 64,
        key: (+new Date()).toString(),
        action,
        cancelAction,
        type: "error",
      },
    ]);
  };
  const success = (toast: Partial<IToast> | string) => {
    let text, preserve, action, cancelAction;
    if (typeof toast === "string") {
      text = toast;
    } else {
      ({ text, preserve, action, cancelAction } = toast);
    }
    setMessages((prev) => [
      ...prev,
      {
        text,
        preserve,
        height: action || cancelAction ? 72 : 64,
        key: (+new Date()).toString(),
        action,
        cancelAction,
        type: "success",
      },
    ]);
  };
  const removeToastByKey = (key: string) => {
    setMessages((prev) => prev.filter((p) => p.key !== key));
  };

  return (
    <ToastsContext.Provider
      value={{
        current: {
          messages,
          message,
          error,
          success,
          removeToastByKey,
        },
      }}
    >
      {children}
    </ToastsContext.Provider>
  );
};

export default ToastsProvider;
