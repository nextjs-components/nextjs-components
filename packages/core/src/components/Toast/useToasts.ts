import { useContext } from "react";

import { ToastsContext } from "./ToastsProvider";

export default function useToasts() {
  return useContext(ToastsContext);
}
