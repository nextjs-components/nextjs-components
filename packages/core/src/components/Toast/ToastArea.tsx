import { Portal } from "@radix-ui/react-portal";
import clsx from "clsx";
import React from "react";
import { useReducer, useRef } from "react";
import { useHover } from "react-aria";

import ToastContainer from "./ToastContainer";
import styles from "./toasts.module.css";
import useToasts from "./useToasts";

function useForceUpdate(): () => void {
  return useReducer(() => ({}), {})[1];
}

const ToastArea = () => {
  const { hoverProps, isHovered } = useHover({});
  const { current } = useToasts();

  const messages = current?.messages ?? [];

  const heights = useRef([]);

  const forceUpdate = useForceUpdate();

  return (
    <Portal>
      <div
        {...hoverProps}
        className={clsx("toast-area", styles.toastArea, {
          [styles.multiple]: messages.length > 1,
        })}
      >
        {messages.map((e, i) => {
          return (
            <ToastContainer
              key={e.key}
              hovering={isHovered}
              position={messages.length - 1 - i}
              text={e.text}
              height={e.height}
              heights={heights.current}
              remove={() => {
                current?.removeToastByKey(e.key);
              }}
              type={e.type}
              preserve={e.preserve}
              action={e.action}
              cancelAction={e.cancelAction}
              id={e.key}
              onMount={forceUpdate}
            />
          );
        })}
      </div>
    </Portal>
  );
};

export default ToastArea;
