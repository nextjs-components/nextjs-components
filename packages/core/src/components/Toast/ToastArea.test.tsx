import { act, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useHover } from "react-aria";
import { mocked } from "ts-jest/utils";

import ToastArea from "./ToastArea";
import ToastContainer from "./ToastContainer";
import { IToastsContext } from "./ToastsProvider";
import useToasts from "./useToasts";

jest.mock("./useToasts");
const useToastsMock = mocked(useToasts);

jest.mock("./ToastContainer");
const ToastContainerMock = mocked(ToastContainer);

jest.mock("react-aria");
const useHoverMock = mocked(useHover);

jest.useFakeTimers();
jest.mock("components/Button", () => ({
  __esModule: true,
  Button: (props) => <button {...props} />,
}));

const ToastContainerFn = jest.fn(() => <div>test</div>);

describe("ToastArea", () => {
  beforeEach(() => {
    // TODO This appears that it should work, but doesn't
    // - useHover.isHovered should be `false` but is `undefined`
    // useHoverMock.mockImplementation(() => {
    //   const originalModule = jest.requireActual("react-aria");
    //   return originalModule.useHover;
    // });
    useHoverMock.mockImplementation(() => ({
      hoverProps: {},
      isHovered: false,
    }));
    useToastsMock.mockImplementation(() => jest.requireActual("./useToasts"));
    ToastContainerMock.mockImplementation(ToastContainerFn as any);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  // GIVEN
  it("should render without any errors", () => {
    // WHEN
    const { baseElement } = render(<ToastArea />);
    // THEN
    expect(baseElement).toMatchSnapshot();
  });

  // GIVEN
  it("should be styled correctly, when there are multiple messages", () => {
    // WHEN
    useToastsMock.mockImplementationOnce(() => {
      return {
        current: { messages: [{ key: 1 }, { key: 2 }] },
      } as unknown as IToastsContext;
    });

    const { baseElement } = render(<ToastArea />);
    const toastArea = baseElement.getElementsByClassName("toast-area")[0];

    // THEN
    expect(toastArea).toHaveClass("multiple");
  });

  // GIVEN
  it("should render ToastContainer children in the correct positions", () => {
    // WHEN
    useToastsMock.mockImplementation(() => {
      return {
        current: { messages: [{ key: 1 }, { key: 2 }, { key: 3 }] },
      } as unknown as IToastsContext;
    });

    render(<ToastArea />);

    // THEN
    expect(ToastContainerFn).toBeCalledTimes(3);

    // @ts-ignore
    expect(ToastContainerFn.mock.calls[2][0]).toMatchObject({
      hovering: false,
      id: 3,
      position: 0,
    });
    // @ts-ignore
    expect(ToastContainerFn.mock.calls[0][0]).toMatchObject({
      hovering: false,
      id: 1,
      position: 2,
    });
  });

  // GIVEN;
  describe("hovering", () => {
    it("should result in ToastContainer children having hovered styles", () => {
      useHoverMock.mockImplementation(() => ({
        hoverProps: {},
        isHovered: true,
      }));
      useToastsMock.mockImplementation(() => {
        return {
          current: { messages: [{ key: 1 }] },
        } as unknown as IToastsContext;
      });

      const { container } = render(<ToastArea />);

      act(() => {
        userEvent.hover(container);
      });

      // @ts-ignore
      expect(ToastContainerFn.mock.calls[0][0]).toMatchObject({
        hovering: true,
      });
    });
  });
});
