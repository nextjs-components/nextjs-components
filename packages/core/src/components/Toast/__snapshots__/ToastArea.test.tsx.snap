// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastArea should render without any errors 1`] = `
<body>
  <div />
  <div>
    <div
      class="toast-area toastArea"
    />
  </div>
</body>
`;
